### PR TITLE
collaboration endpoint

### PR DESCRIPTION
**Pull request by carycheng**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @carycheng, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: boxsdk/client/client.py [hunk 1]
```diff
@@ -524,6 +524,37 @@ def create_user(self, name, login=None, **user_attributes):
...
+        :type fields:
+            `Iterable` of `unicode`
+        :returns:
+            An iterator of the entries in the pending collaborations
+        :rtype:
+            :class:`BoxObjectCollection`
+        """
+        return LimitOffsetBasedObjectCollection(
+            session=self._session,
+            url=self.get_url('collaboration'),
```
Comment (mattwiller) [inline boxsdk/client/client.py]: This URL doesn't look correct; I think it should be `self.get_url('collaborations')`.  On a related note, you're missing a test for this method.

File: boxsdk/object/collaboration.py [hunk 2]
```diff
@@ -59,3 +59,13 @@ def update_info(self, role=None, status=None):
         if status:
             data['status'] = status
         return super(Collaboration, self).update_info(data=data)
+
+    def accept(self):
+        """Accepts a pending collaboration"""
+        data = {'status': 'accepted'}
+        return super(Collaboration, self).update_info(data=data)
```
Comment (mattwiller) [inline boxsdk/object/collaboration.py]: You could just call the method above these, I think.  For example:

```python
return self.update_info(status='accepted')
```

File: boxsdk/object/item.py [hunk 3]
```diff
@@ -398,3 +399,133 @@ def remove_from_collection(self, collection):
...
+            },
+            'role': role,
+        }
+        if can_view_path is not None:
+            body['can_view_path'] = can_view_path
+        params = {}
+        if fields is not None:
+            params['fields'] = ','.join(fields)
+        if notify is not None:
+            params['notify'] = ','.join(fields)
```
Comment (mattwiller) [inline boxsdk/object/item.py]: This looks like an incorrect copy/paste; should probably be just `params['notify'] = notify`

File: boxsdk/object/item.py [hunk 4]
```diff
@@ -398,3 +399,133 @@ def remove_from_collection(self, collection):
...
+            `bool` or None
+        :param notify:
+            Determines if the collaborator should receive a notification for the collaboration.
+        :type notify:
+            `bool` or None
+        :param fields:
+            List of fields to request.
+        :type fields:
+            `Iterable` of `unicode`
+        """
```
Comment (mattwiller) [inline boxsdk/object/item.py]: This docstring is missing return value information

File: boxsdk/object/item.py [hunk 5]
```diff
@@ -398,3 +399,133 @@ def remove_from_collection(self, collection):
...
```
Comment (mattwiller) [inline boxsdk/object/item.py]: This docstring is missing return value information

File: boxsdk/object/item.py [hunk 6]
```diff
@@ -398,3 +399,133 @@ def remove_from_collection(self, collection):
...
+            },
+            'role': role,
+        }
+        if can_view_path is not None:
+            body['can_view_path'] = can_view_path
+        params = {}
+        if fields is not None:
+            params['fields'] = ','.join(fields)
+        if notify is not None:
+            params['notify'] = ','.join(fields)
```
Comment (mattwiller) [inline boxsdk/object/item.py]: Incorrect copy/paste, not attaching the correct value

File: boxsdk/object/item.py [hunk 7]
```diff
@@ -398,3 +399,133 @@ def remove_from_collection(self, collection):
...
+        response = self._session.post(url, data=json.dumps(body), params=params).json()
+        return self.translator.translate(response['type'])(
+            session=self._session,
+            object_id=response['id'],
+            response_object=response,
+        )
+
+    def collaborations(self, limit=None, marker=None, fields=None):
+        """
+        Get the entries in the collaboration using marker-based paging.
```
Comment (mattwiller) [inline boxsdk/object/item.py]: We can just remove the part about marker-based paging form this description; that's an implementation detail that the user probably doesn't care about.

File: docs/collaboration.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+collaborator = client.user(user_id)
+collaboration = client.folder(folder_id).collaborate(collaborator, CollaborationRole.VIEWER)
+```
+
+Alternatively you can also invite a user with their email address.
+
+```python
+from boxsdk.object.collaboration import CollaborationRole
+folder_id = '1234'
+email_of_invitee = "testuser@example.com"
```
Comment (mattwiller) [inline docs/collaboration.md]: String quoting should be consistent between this line and the one above (I think we generally prefer single quotes)

File: docs/collaboration.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+Edit a Collaboration
+--------------------
+
+A collaboration can be edited by using `collaboration.update_info()`.
+
+```python
+from boxsdk.object.collaboration import CollaborationRole
+collaboration_id = '1111'
+collaboration_update = {'role': CollaborationRole.EDITOR}
+updated_collaboration = client.collaboration(collaboration_id).update_info(collaboration_update)
```
Comment (mattwiller) [inline docs/collaboration.md]: This example doesn't use the overridden version of this method that takes explicit `status` and `role` params; should we update that method and use it in these examples?

File: docs/collaboration.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+```
+
+List Pending Collaborations
+------------------------------------
+
+To retrieve all pending collaborations for the user, use `folder.pending_collaborations(limit=None, offset=None, fields=None)`.
+
+```python
+pending_collaborations = client.pending_collaborations()
+for pending_collaborations in pending_collaborations:
```
Comment (mattwiller) [inline docs/collaboration.md]: I don't think this example will work; you're using the same variable name (`pending_collaborations`) twice

File: docs/collaboration.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+Get Collaborations on a File
+----------------------------
+
+To retrieve an iterable of collaborations on a specified file you can use `file.collaborations(limit=None, marker=None, fields=None)`.
+
+```python
+file_id = '2222'
+collaborations = client.file(file_id).collaborations()
+for collaborations in collaborations_on_file:
```
Comment (mattwiller) [inline docs/collaboration.md]: This example is broken; you're defining one variable name (`collaborations`) but then using a different one to iterate over (`collaborations_on_file`)

File: docs/collaboration.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+List Collaborations on a Folder
+-------------------------------
+
+To retrieve all collaborations on a specified folder you can use `folder.collaborations(limit=None, marker=None, fields=None)`.
+
+```python
+folder_id = '1234'
+collaborations = client.folder(folder_id).collaborations()
+for collaborations in collaborations_on_folder:
```
Comment (mattwiller) [inline docs/collaboration.md]: This example is broken; you're defining one variable name (`collaborations`) but then using a different one to iterate over (`collaborations_on_folder`)

File: docs/collaboration.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+```
+
+Accept or Reject a Pending Collaboration
+-----------------------------------------
+
+To `accept` or `reject` a pending collaboration use `collaboration.accept()` or `collaboration.reject()`.
+
+```python
+collaboration_id = '3333'
+client.collaboration(collaboration_id).accept()
```
Comment (mattwiller) [inline docs/collaboration.md]: You might want to indicate the return value for this method (it's the updated collaboration object)

File: docs/collaboration.md [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+```
+
+You can reject a pending collaboration with
+
+```python
+collaboration_id = '3333'
+client.collaboration(collaboration_id).reject()
```
Comment (mattwiller) [inline docs/collaboration.md]: Should probably indicate the return value here too

File: test/unit/object/test_collaboration.py [hunk 15]
```diff
@@ -34,3 +34,51 @@ def test_update_info_returns_the_correct_response(
...
+        params=None,
+    )
+    assert isinstance(response, test_collaboration.__class__)
+    assert response.status == new_status
+
+
+def test_reject_pending_collaboration(test_collaboration, mock_box_session):
+    # pylint:disable=protected-access
+    new_status = 'rejected'
+    expected_url = test_collaboration.get_url()
```
Comment (mattwiller) [inline test/unit/object/test_collaboration.py]: I'd prefer a more explicit check for these URLs; what if `get_url()` is wrong?

File: boxsdk/object/item.py [hunk 16]
```diff
@@ -398,3 +399,141 @@ def remove_from_collection(self, collection):
...
+        if notify is not None:
+            params['notify'] = notify
+        response = self._session.post(url, data=json.dumps(body), params=params).json()
+        return self.translator.translate(response['type'])(
+            session=self._session,
+            object_id=response['id'],
+            response_object=response,
+        )
+
+    def collaborations(self, limit=None, marker=None, fields=None):
```
Comment (mattwiller) [inline boxsdk/object/item.py]: Should probably be `get_collaborations()`

File: docs/collaboration.md [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+To retrieve all collaborations on a specified folder you can use `folder.collaborations(limit=None, marker=None, fields=None)`.
+
+```python
+folder_id = '1234'
+collaborations = client.folder(folder_id).collaborations()
+for collaboration in collaborations:
+    # Do something
+```
+
+Get Collaborations on a File
```
Comment (mattwiller) [inline docs/collaboration.md]: You can probably condense this with the section above

File: docs/collaboration.md [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+file_id = '2222'
+collaborations = client.file(file_id).collaborations()
+for collaboration in collaborations
+    # Do something
+```
+
+List Pending Collaborations
+------------------------------------
+
+To retrieve all pending collaborations for the user, use `folder.pending_collaborations(limit=None, offset=None, fields=None)`.
```
Comment (mattwiller) [inline docs/collaboration.md]: This signature should be updated to reflect that the method is on `Client`, not `Folder`

File: test/unit/client/test_client.py [hunk 19]
```diff
@@ -387,6 +388,31 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
+        'limit': 2,
+        'offset': 0,
+        'entries': [mock_collaboration],
+    }
+    pending_collaborations = mock_client.pending_collaborations(limit=2)
+    pending_collaboration = pending_collaborations.next()
+    mock_box_session.get.assert_called_once_with(expected_url, params={'limit': 2, 'status': 'pending', 'offset': None})
+    assert isinstance(pending_collaboration, Collaboration)
+    assert pending_collaboration.id == mock_collaboration['id']
+    assert pending_collaboration.type == mock_collaboration['type']
```
Comment (mattwiller) [inline test/unit/client/test_client.py]: Should also assert about `created_by`

File: test/unit/object/test_item.py [hunk 20]
```diff
@@ -215,3 +218,120 @@ def test_remove_from_collection(test_item_and_response, mock_box_session, mock_c
...
+    mock_box_session.post.assert_called_once_with(expected_url, data=json.dumps(expected_data), params={})
+    assert collaboration.id == mock_collaboration['id']
+    assert collaboration['type'] == mock_collaboration['type']
+    assert collaboration['created_by']['id'] == mock_collaboration['created_by']['id']
+
+
+def test_collaborate_with_login(test_item_and_response, mock_box_session):
+    # pylint:disable=redefined-outer-name, protected-access
+    test_item, _ = test_item_and_response
+    expected_url = mock_box_session.get_url('collaborations')
```
Comment (mattwiller) [inline test/unit/object/test_item.py]: Expected URL should be manually constructed to ensure correctness

File: test/unit/object/test_item.py [hunk 21]
```diff
@@ -215,3 +218,120 @@ def test_remove_from_collection(test_item_and_response, mock_box_session, mock_c
...
+    mock_box_session.post.assert_called_once_with(expected_url, data=json.dumps(expected_data), params={})
+    assert collaboration.id == mock_collaboration['id']
+    assert collaboration['type'] == mock_collaboration['type']
+    assert collaboration['created_by']['id'] == mock_collaboration['created_by']['id']
+
+
+def test_collaborations(test_item_and_response, mock_box_session):
+    # pylint:disable=redefined-outer-name, protected-access
+    test_item, _ = test_item_and_response
+    expected_url = test_item.get_url('collaborations')
```
Comment (mattwiller) [inline test/unit/object/test_item.py]: Expected URL should be manually constructed to ensure correctness

File: test/unit/object/test_item.py [hunk 22]
```diff
@@ -215,3 +218,120 @@ def test_remove_from_collection(test_item_and_response, mock_box_session, mock_c
...
+    mock_box_session.get.return_value.json.return_value = {
+        'limit': 500,
+        'entries': [mock_collaboration]
+    }
+    collaborations = test_item.collaborations(limit=500)
+    collaboration = collaborations.next()
+    mock_box_session.get.assert_called_once_with(expected_url, params={'limit': 500})
+    assert isinstance(collaboration, Collaboration)
+    assert collaboration.id == mock_collaboration['id']
+    assert collaboration.type == mock_collaboration['type']
```
Comment (mattwiller) [inline test/unit/object/test_item.py]: Should also assert about `created_by` here

File: docs/collaboration.md [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+```python
+collaboration_id = '1111'
+collaboration_info = client.collaboration(collaboration_id).get()
+```
+
+List Collaborations on a Folder
+-------------------------------
+
+To retrieve all collaborations on a specified folder you can use `folder.collaborations(limit=None, marker=None, fields=None)`.
```
Comment (mattwiller) [inline docs/collaboration.md]: Update to `get_collaborations()` along with the code

**Pull request merged**